### PR TITLE
Remove coef_ requirement from ONNX exporter.

### PR DESCRIPTION
**Pull request by bogedy**
This PR is my attempt to solve the todo to make the ONNX exporter work with any sklearn `model_head` that doesn't have the `coef_` attribute, which was deprecated for `OneVsRestClassifier` in sklearn 1.1. 

I just use `n_features_in_` for the shape and for the dtype I check if it has a `coef_` attribute and if not I take the `coef_` attribute of the first element of `estimators_`.

I also expanded the ONNX test file script.

This is my first PR on anything, my apologies if I misunderstood the scope of the problem or did anything else wrong!

### CODE REVIEW

File: src/setfit/exporters/onnx.py [hunk 1]
```diff
@@ -136,15 +136,12 @@ def export_sklearn_head_to_onnx(model_head: LogisticRegression, opset: int) -> o
...
-
     # Determine the initial type and the shape of the output.
-    input_shape = (None, *model_head.coef_.shape[1:])
-    dtype = guess_data_type(model_head.coef_, shape=input_shape)[0][1]
+    input_shape = (None, model_head.n_features_in_)
+    if hasattr(model_head, "coef_"):
+        dtype = guess_data_type(model_head.coef_, shape=input_shape)[0][1]
+    if not hasattr(model_head, "coef_") and hasattr(model_head, "estimators_"):
+        dtype = guess_data_type(model_head.estimators_[0].coef_, shape=input_shape)[0][1]
     dtype.shape = input_shape
```
Comment (tomaarsen) [inline src/setfit/exporters/onnx.py:145]: Isn't there an edge case now where `dtype` is undefined?

File: tests/exporters/test_onnx.py [hunk 2]
```diff
@@ -11,9 +11,24 @@
...
+    [
+        ("lewtun/my-awesome-setfit-model", ["i loved the spiderman movie!", "pineapple on pizza is the worst 🤮"]),
+        (
+            "lewtun/setfit-ethos-multilabel-example",
+            [
+                "Women are made for making babies and cooking dinner and nothing else!!!",
+                "Lmfao that fat slob in the wheelchair",
+                "Indian Ned Flanders. You look like you professionally harrass subpar women on Facebook",
+                "Aren’t you ashamed about your body colour?",
+                "Islam is  the home of terriorism .",
```
Comment (tomaarsen) [inline tests/exporters/test_onnx.py:25]: I understand that the ethos model is a hate-speech detection model, but because we're not actually checking the correctness of the model outputs, I think I would prefer more "normal" sentences here, if that's okay with you :)

File: tests/exporters/test_onnx.py [hunk 3]
```diff
@@ -48,6 +62,9 @@ def test_export_onnx_sklearn_head():
         # Compare the results and ensure that we get the same predictions.
         assert np.array_equal(onnx_preds, pytorch_preds)
 
+    except Exception as e:
+        raise e
+
```
Comment (tomaarsen) [inline tests/exporters/test_onnx.py:67]:
```suggestion
```
I think this is equivalent to not writing this. If there's an error in this try-finally, it'll get raised even if this `except` is not here.

**Review by tomaarsen — COMMENTED** [review]
This is looking good! Especially for a first PR. However, I do have some notes/questions.
First of all, could you run `make style` to run some formatting rules, and then push those changes? This will satisfy the currently failing quality checks.

**Comment by bogedy** [comment]
Sorry, didn't not realize that pushing to my own branch would update here. First PR...

File: src/setfit/exporters/onnx.py (hunk 1 above)
Comment (bogedy) [reply src/setfit/exporters/onnx.py:145]: It probably suffices to only support models with either `coef_` or where every item in `estimators_` has a `coef_`.

File: tests/exporters/test_onnx.py (hunk 2 above)
Comment (tomaarsen) [reply tests/exporters/test_onnx.py]: I'm fairly certain that difference is considerably larger than it should be, indeed. I'd love for you to try and help debug it!

**Comment by tomaarsen** [comment]
No worries! It'll all get squashed into one neat commit when I merge this, so you don't have to worry too much about keeping the PR "tidy" with few commits, etc. :)

File: tests/exporters/test_onnx.py (hunk 2 above)
Comment (bogedy) [reply tests/exporters/test_onnx.py]: It might just be this old, unsolved problem https://github.com/huggingface/setfit/issues/291#issue-1562241805. It's probably because `lewtun/my-awesome-setfit-model` uses mpnet.

**Comment by bogedy** [comment]
Don't know why the quality check isn't passing. I ran `make style && make quality` repeatedly and double checked that I had the right versions installed. I followed all the dev instructions.

**Comment by tomaarsen** [comment]
@bogedy 

I ran the `style` locally, as I think my local black & isort versions align with the ones that the CI uses. 
Do you think the PR is ready?

**Comment by bogedy** [comment]
Hi! I'm sorry I fell off this, got really busy. The PR is ready. I thought maybe I should upload a new model to hugging face since the lewtun my awesome model appears to be bugged with ONNX, but I suppose that can be taken care of later.

**Comment by bogedy** [comment]
I thought this needed a lot more and neglected it but looking at the error in the unit test it was just random chance? 

502 Server Error: Bad Gateway for url: https://huggingface.co/lewtun/my-awesome-setfit-model/resolve/a3c0fe9302019e63581ad34ab25e6b6c5b4aab5d/pytorch_model.bin

That URL works for me...

**Comment by tomaarsen** [comment]
I get similar test failures here: #397. I haven't figured out how these started happening, as the Openvino version hasn't changed to my knowledge.

**Comment by bogedy** [comment]
I found the issue I think. It's hummingbird_ml-0.4.9. Downgrading to 0.4.8 makes the tests pass for me.

**Comment by tomaarsen** [comment]
I'll merge main into this so we can rerun the tests :)

**Comment by tomaarsen** [comment]
This is looking good I think! Thanks, and apologies for the delays and random CI issues. Oh, and thanks for fixing the latest CI issues for me!

**Comment by bogedy** [comment]
@tomaarsen thank you so much for holding my hand through my first ever PR, sorry I neglected it for months haha

**Comment by tomaarsen** [comment]
Haha no worries, it's also my bad for failing to review this PR for weeks or months. No big deal all around
